### ISSUE TITLE
slow download speed PIA/open ports question about configuration

### ISSUE THREAD

**Issue by sesshomaru86** [haugene/docker-transmission-openvpn#625]
Hello ! First of all thanx for this amazingly docker container ! After i changed my ddwrt router for a unifi one i couldn't run anymore the vpn on my nas and as it doesn't support docker i install it on external linux server and really it works amazingly. I have only one problem , download speed. I know that it's a topic that was opened multiple times and closed too but after reading every topic multiple times , couldn't still figure out a flawless performance so i supposed it wasnt a better place to ask than the place where the project is. I Know that it's not a bug just a misconfiguration of mine probably but i thought that people with similar configuration can help me out. I'm using Pia as vpn provider and a port forwarding compatible server so as i read from the logs and then from the interface the port is open so it should be any problem . Normally with  the Pia app in linux as also in windows desktop linux images are downloaded with over 10Mb/s , same speed that i had configuring the openvpn client on a linksys router with ddwrt. Now through the docker speeds aren't more than 1mbs. In my docker composer file the only port that is specified is the 9091 that is the web interface. Should i put also the port provided by pia ? i thought that is assigned automatically by the the vpn provider through the vpn tunnel. Also do i need to open from my router too ? If not some suggestion of other things that i can do maybe directly on docker environment ? I'm using also pihole with dns over https . Can somehow affect the correct behaviour of the container even if i think it take direclty the dns from the compose file that are specified the cloudflare ones. Really apprciate any input.

Here is a copy of my compose file

 container_name: transmission-vpn
     image: haugene/transmission-openvpn
     cap_add:
       - NET_ADMIN
     devices:
       - /dev/net/tun
     restart: always
     ports:
     - "9091:9091"
     dns:
       - 1.1.1.1
       - 1.0.0.1
     volumes:
       - /etc/localtime:/etc/localtime:ro
       - ${USERDIR}/docker/transmission-vpn:/data
       - ${USERDIR}/docker/shared:/shared
       - ${USERDIR}/download/watch:/data/watch
       - ${USERDIR}/download/completed:/data/completed
       - ${USERDIR}/download/incomplete:/data/incomplete
     environment:
       - OPENVPN_PROVIDER=PIA
       - OPENVPN_USERNAME=******
       - OPENVPN_PASSWORD=*******
       - OPENVPN_CONFIG="DE Berlin"
       - OPENVPN_OPTS=--inactive 3600 --ping 10 --ping-exit 60
       - LOCAL_NETWORK=192.168.5.0/24
       - PUID=${PUID}
       - PGID=${PGID}
       - TZ=${TZ}
       - TRANSMISSION_RPC_AUTHENTICATION_REQUIRED=true
       - TRANSMISSION_RPC_HOST_WHITELIST="127.0.0.1,192.168.*.*"
       - TRANSMISSION_RPC_PASSWORD=*******
       - TRANSMISSION_RPC_USERNAME=sessho
       - TRANSMISSION_UMASK=002
       - TRANSMISSION_RATIO_LIMIT=1.00
       - TRANSMISSION_RATIO_LIMIT_ENABLED=true
       - TRANSMISSION_SPEED_LIMIT_UP=10
       - TRANSMISSION_SPEED_LIMIT_UP_ENABLED=true
       - TRANSMISSION_BLOCKLIST_ENABLED=true
       - TRANSMISSION_BLOCKLIST_URL=http://john.bitsurge.net/public/biglist.p2p.gz

**Comment by alneven** [comment]
what I have seen, that my upload speed via PIA VPN is rellay bad compared to the networks speeds I have without VPN
but it doesn't matter

**Comment by sesshomaru86** [comment]
Hmm nada... Tried changing dns with the one that pia provides , changed the country to the one that im currently living the speed doesn't increase a bit..Ram and cpu usage isn't set in limiting resources so it shouldn't be  a problem . I recreate the container to refresh all the settings but still nothing.At that point i'm out of ideas. I will try it maybe to another machine or wait for a newer version hoping it will resolve this problem.

**Comment by haugene** [comment]
I'll post the same answer here as I just made in the gitter chat. Maybe someone can do some digging :)
**TL;DR:** This is probably not an issue with the image, but some kind of local network or docker stuff.

I just spun up a VM on DigitalOcean, installed Docker and then started this container using a very basic setup:
```
#!/usr/bin/env bash

docker run --name transmission_openvpn \
    --cap-add=NET_ADMIN --device=/dev/net/tun -d \
    --dns 8.8.8.8 --dns 8.8.4.4 \
    -v /root/torrent/:/data \
    -v /etc/localtime:/etc/localtime:ro \
    -e OPENVPN_PROVIDER=PIA \
    -e OPENVPN_USERNAME=AzureDiamond \
    -e OPENVPN_PASSWORD=hunter2 \
    --log-driver json-file \
    --log-opt max-size=10m \
    -p 9091:9091 \
    haugene/transmission-openvpn

docker run -d \
    --name proxy \
    --link transmission_openvpn:transmission \
    -p 8080:8080 \
    haugene/transmission-openvpn-proxy
```
I then added the ubuntu 18.10 server torrent and got speeds of 10MB/s+. So I feel that verifies that this is working properly. Any other results is therefore not related to just docker, openvpn and transmission - but how it is set up locally.

You can try this yourself, use this referral link https://m.do.co/c/ca994f1552bc and you'll get 100$ credit on DigitalOcean now in October (usually 25$).

To resolve your issue I refer to my previous answer. Test DL speeds in a plain ubuntu image, then in the vpn container (not through torrents) and maybe in a torrent conainer without VPN and see where the troubles lie for you.

**Comment by rcdailey** [comment]
@haugene 

* What's the proxy container for?
* What is the default `OPENVPN_CONFIG` for PIA? If it's not one of the servers supporting port forwarding, how do you accept peers for torrents? Or is that not relevant to this test?

**Comment by haugene** [comment]
* The proxy is proxying my web traffic, so I can access the UI. 
When starting it up in the cloud I guess I could set LOCAL_NETWORK to be my home public IP, but that's not exactly what I want (I want to be able to access it anywhere) and my public IP is not exactly a "local network". 
To be able to connect to my UI while on mobile networks etc I would basically have to set LOCAL_NETWORK to a wide CIDR address - and that's definitely not what I want. That will allow traffic outside the VPN connection to a large range of hosts, effectively making my torrents insecure.
So I just want to forward my web port - and that's why I use the proxy. It is on the Docker network, so it's on the openvpn-container local network by default, and only forwards traffic on my web port so all the other traffic is still going through VPN. This is described in the README and in several issues where people have questions about the proxy.

* As for the OPENVPN_CONFIG, when not set it will go to the default server which for PIA is [CA Toronto](https://github.com/haugene/docker-transmission-openvpn/blob/master/openvpn/pia/default.ovpn) and that [supports port forwarding](https://www.privateinternetaccess.com/helpdesk/kb/articles/how-do-i-enable-port-forwarding-on-my-vpn). So, all good.

**Comment by rcdailey** [comment]
@haugene Thanks for explaining. So you are saying you don't want to set `LOCAL_NETWORK` to your WAN IP, right? I guess that means you don't have a LAN or a NAT? I have mine set to `LOCAL_NETWORK=192.168.1.0/24` to allow LAN clients access to `9091`, and for WAN access I use a traefik reverse proxy. So it's pretty safe, assuming clients on my LAN are trusted (and they are).

Is this not your set up? Is there any reason to use the proxy in my scenario for less scope of access? In your case, I'd be using a double reverse proxy:

    WAN CLIENT -> ROUTER -> TRAEFIK (443) -> TRANSMISSION PROXY (8080) -> TRANSMISSION (9091)

This seems a bit overkill unless I'm missing something.

Also in your case, you probably shouldn't be mapping port `9091` to the host since you access it through the proxy's port `8080` right? Assuming all you need is to link the containers, which you do, that traffic shouldn't be going through the host itself.

**Comment by haugene** [comment]
Yes. Running on a cloud server I wouldn't set local network to gain access from my home network. Then I would use a proxy instead.

And this was just a very quick example to test the download speed, that was the issue you had. But no. Port 9091 could/should not be mapped to the host 

On a typical setup in a home network on a server or nas, behind nat, the local network variable comes in handy when hosts on the network are trusted as you say.

An no. If you already have a reverse proxy, you don't need one more. This is just an example proxy to get people started. I'm guessing you are running traefik with https so that's a better solution. 

There are a couple of scenarios where the dockerized proxy is still handy and that is mostly when the containers are running in some sort of vm on the host (docker for Mac, windows, some nas servers). Then you might have several layers of networking and nat'ing making the local network variable more tricky to get right. Then a sibling container is handy to break out of the docker network. If traefik is running in a container on the same docker network and is proxying to the openvpn container through that network - then it's again the same.

**Comment by rcdailey** [comment]
Thanks once again. Sorry to go slightly off topic, but in the end it was worth it for the abundant information :-)

**Comment by haugene** [comment]
@brianmorton I appreciate the heads-up. Could have been bad, but this particular username/password is an old classic ;)

http://bash.org/?244321
https://knowyourmeme.com/memes/hunter2

**Comment by cwiggs** [comment]
I'm having a similar issue and when I run a speedtest from inside the container I seem to get good speeds.

```
./speedtest-cli 
Retrieving speedtest.net configuration...
Testing from M247 Europe SRL (185.230.127.241)...
Retrieving speedtest.net server list...
Selecting best server based on ping...
Hosted by Vodafone Kabel Deutschland (Berlin) [1.85 km]: 21.819 ms
Testing download speed................................................................................
Download: 340.17 Mbit/s
Testing upload speed................................................................................................
Upload: 177.14 Mbit/s
```

am i correct to assume that the speedtest will go through the VPN if I run the speedtest inside the container.

**Comment by brianmorton** [comment]
@cwiggs Yes I believe that is correct.

**Comment by haugene** [comment]
Yes @cwiggs. As long as the IP the test is run against (185.230.127.241) is outside what you've defined as "LOCAL_NETWORK" the traffic will go through VPN. That IP should definitely be outside the range.

So then the VPN performs, the container performs - but Transmission is slow? Sounds like a Transmission or tracker related problem. Maybe volume driver for Docker in extreme cases, but that slow...?

**Comment by cwiggs** [comment]
I think my issue is boiling down to a tracker(s) issue, not with this docker image.  Thanks for the reponse @haugene and thanks for providing this awesome image!
